oeps/oep-0051-bp-conventional-commits.rst

```diff
@@ -0,0 +1,184 @@
+
+The nearer to the change that information is captured, the easier it will be to write, and the more accurate it will be.  Changes made by developers are consumed by many people, from the pull request reviewer to the final user of the product.  The more the developer can explain the change, the more easily the change will be understood and used, and the more efficient the overall development process will be.
+
+Commit messages are the closest text to the change itself.  The more
+information-rich we can make commit messages, the more useful they will be.
+
+
+Specification
+============
+
+We are adopting the `Conventional Commits`_ spec, with our own adjustments.
+
+.. note::
+   Many people are familiar with the `Angular commit message format`_, which uses conventional commits and inspired the Conventional Commits spec.  We are **not** adopting the Angular rules.
+
+Git commit messages have these parts::
+
+    <type>: <subject>
+
+    <body>
+
+    <footer>
+
+Type
+----
+
+We use these commit types:
+
+* **build**: anything to do with building or releasing the repo: Makefile, tox.ini, Travis, Jenkins, GitHub actions, and so on.
+
+* **chore**: a repetitive mechanical task such as updating requirements or translations.
+
+* **docs**: documentation-only changes. Documentation changes for other types of work should  be included in those commits. This includes more than the formal docs for a repo, it also covers READMEs, ADRs, docstrings, and so on.
+
+* **feat**: a new feature, or a change to an existing feature. Anything that changes the set of features.  Public API entry points and their behavior are part of the feature set of a product, so changes to those should be marked with "feat".
+
+* **fix**: bug fixes, including changes to the behavior of features, and changes that remove or mitigate security vulnerabilities.
+
+* **perf**: performance improvement.
+
+* **refactor**: code reorganization that doesn't change behavior from the repo consumer perspective.
+
+* **revert**: undo a previous change. The previous commit message, including the type prefix, is included.
+
+* **style**: improve the styling of the code.
+
+* **test**: test-only changes. Adds missing tests or corrects existing tests. Tests accompanying other types of changes go into those commits.
+
+* **exp**: experimental or exploratory changes that are not meant to be permanent.  Occasionally changes have to be pushed to GitHub or even production that are intended to be short-lived. For example, debugging continuous integration, or ad-hoc debugging on live systems.
+
+
+Breaking changes include an exclamation point as part of the type::
+
+    feat!: remove the ability to author courses
+
+Ideally, a single commit won't mix these types together.  If a commit does, use the most important type in the commit message.  The priority order for the types is generally: revert, feat, fix, perf, docs, test, build, refactor, style, chore, exp.
```

Comment:
I don't know how important it is to keep every commit pure by type.  If I make a commit that updates a large comment, and nothing else, that's a "docs" change.  But that doesn't mean that all changes to comments should be in a separated "docs" commit.  How bad is it if a feature commit also includes a fix?

(This is why I want to eventually get to commits including updates to changelogs: it gives us the space to fully describe what is happening.)

Reply:
If I implement a feature and write docs for that feature, should they be in two separate commits?  I don't understand what value it would add to split them up like that?

Reply:
That's totally fair - especially regarding docs.

In my mind, I was thinking about keeping the following types of commits separate (as examples):

- `refactor` commits separate from `feat` commits - so it's easier for consumers to understand what functionality actually changed.
- 'style' commits separate - so consumers can easily validate those changes in a separate commit and spend more time on other substantial changes.
- 'temp' commits separate - allows easier removal of those changes in the future.

This suggestion is not a blocker for this OEP.  But a suggestion of this sort can help authors improve the presentation of their code changes and can help consumers better understand the changes.